libs/ballot-interpreter/src/hmpb-rust/timing_marks/corners/shape_finding/mod.rs

```diff
@@ -0,0 +1,259 @@
+            // as good reference points for bubble finding, so we'll want to
+            // reject the ballot anyway.
+            if (bounds.left() == image_bounds.left() && bounds.top() == image_bounds.top())
+                || (bounds.left() == image_bounds.left()
+                    && bounds.bottom() == image_bounds.bottom())
+                || (bounds.right() == image_bounds.right() && bounds.top() == image_bounds.top())
+                || (bounds.right() == image_bounds.right()
+                    && bounds.bottom() == image_bounds.bottom())
+            {
+                return None;
+            }
+
+            Some(shape)
+        })
+        .collect()
+}
+
+/// The shape of a timing mark detected by scanning columns.
+#[derive(Debug, Clone, PartialEq)]
+pub struct TimingMarkShape {
+    /// The leftmost x coordinate this timing mark shape contains.
+    pub x: u32,
+
+    /// The ranges of the y coordinates for each of the x coordinates this
+    /// timing mark shape contains, from left to right starting at `self.x`.
+    pub y_ranges: Vec<RangeInclusive<u32>>,
+}
+
+impl TimingMarkShape {
+    /// Iterator for all the points this shape contains. They are all contiguous.
+    pub fn points(&self) -> impl Iterator<Item = Point<PixelUnit>> + '_ {
+        self.y_ranges
+            .iter()
+            .cloned()
+            .zip(self.x..)
+            .flat_map(|(y_range, x)| y_range.map(move |y| Point::new(x, y)))
+    }
+
+    /// The smallest bounding rectangle containing all the points in this shape.
+    pub fn bounds(&self) -> Rect {
+        let (Some(&min_y), Some(&max_y)) = (
+            self.y_ranges.iter().map(RangeInclusive::start).min(),
+            self.y_ranges.iter().map(RangeInclusive::end).max(),
+        ) else {
+            return Rect::zero();
+        };
+
+        let width = self.y_ranges.len() as u32;
+
+        Rect::new(
+            self.x as i32,
+            min_y as i32,
+            width as u32,
+            (max_y - min_y + 1) as u32,
+        )
+    }
+
+    /// Builds a new `TimingMarkShape` with the same range of `x` values but
+    /// with the `y` values smoothed using a median filter whose window size
+    /// scales with the size of the timing mark's width.
+    ///
+    /// See https://en.wikipedia.org/wiki/Median_filter
+    pub fn smoothed(&self, geometry: &Geometry) -> Self {
+        fn median_filter(values: &[u32], window_size: usize) -> Vec<u32> {
+            let half = window_size / 2;
+            (0..values.len())
+                .map(|i| {
+                    let start = i.saturating_sub(half);
+                    let end = (i + half + 1).min(values.len());
+                    let mut window = values[start..end].to_vec();
+                    window.sort();
+                    window[window.len() / 2]
+                })
+                .collect()
+        }
+
+        let window_size = (geometry.timing_mark_size.width / 5.0).ceil() as usize;
```

Comment:
How did you determine a good window size? I don't see much guidance in the wikipedia article other than: 
>Median filtering works well for both positive impulses (spikes) and negative impulses (dropouts), so long as a window can be chosen so that the number of entries infected with impulse noise is (almost) always smaller than half of the window size.

Might be nice to leave a little explanatory comment about choosing this parameter.

I'm also not convinced it needs to scale with the width of the timing mark based on that bit from Wikipedia (though obviously since timing mark width is constant for now, it's a moot point)

Reply:
[23c4d81](https://github.com/votingworks/vxsuite/pull/6582/commits/23c4d8183a9e2857efaf2d3b6cb2ea0074abdc28)